<a href="https://colab.research.google.com/github/RahulMisal161/first/blob/Data-Exploratory-analysis/Copy_of_Virat_ODI.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

---
# **Table of Contents**
---

**1.** [**Introduction**](#Section1)<br>
**2.** [**Installing & Importing Libraries**](#Section3)<br>
  - **2.1** [**Installing Libraries**](#Section31)
  - **2.2** [**Upgrading Libraries**](#Section32)
  - **2.3** [**Importing Libraries**](#Section33)

**3.** [**Data Acquisition & Description**](#Section4)<br>
  - **3.1** [**Data Description**](#Section41)
  - **3.2** [**Data Information**](#Section42)

**4.** [**Data Pre-Profiling**](#Section5)<br>
  - **4.1** [**Pre-profiling report**](#Section51)
 
**5.** [**Data Cleaning**](#Section6)<br>
  - **5.1** [**Handling of Inconsistant Data**](#Section61)
  - **5.2** [**Handling of Missing Values**](#Section611)
  - **5.3** [**Dropping the unnecessary features if any**](#Section62)

**6.** [**Data Post-Profiling**](#Section7)<br>
**7.** [**Exploratory Data Analysis**](#Section8)<br>
  
- **7.1** [**How does the opposition country affect the number of runs virat scores?**](#Section81)
- **7.2** [**What is the reason where he gets out very frequently?**](#Section82)
- **7.3** [**What is the relation between play grounds and runs he scores**](#Section83)
- **7.4** [**What is the relation between the position number at which he play and runs**](#Section84)
- **7.5** [**How much runs he has scored each year and over the years**](#Section85)
- **7.6** [**How does innings(first/second) impacts the how much he scores?**](#Section86)

<br>

**8.** [**Summarization**](#Section9)<br>

  
  
  

  
  
 

---
<a name = Section1></a>
# **1. Introduction**
---

- Everyone knows about **Virat Kohli**, He is considered one of the **greatest batsmen in the world** and has been **ranked as the No. 1 ODI** (One Day International) batsman by the ICC (International Cricket Council) several times..
- He is known for his **aggressive batting style** and his ability to **score runs consistently across all formats** of the game.

- Kohli has also **led the Indian cricket team** to several notable victories, including the **2013 Champions Trophy, the 2016 Asia Cup, and the 2018 Under-19 Cricket World Cup**.

- The sole purpose of this case study is to **analyze** and **identify patterns** of runs he has scored throughout the years in different conditions .


  

---
<a name = Section3></a>
# **2. Installing & Importing Libraries**
---

<a name = Section31></a>
### **2.1 Installing Libraries**

In [1]:
!pip install -q datascience                                         # Package that is required by pandas profiling
!pip install -q pandas-profiling                                    # Library to generate basic statistics about data

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 22.5 MB/s eta 0:00:00


<a name = Section32></a>
### **2.2 Upgrading Libraries**

- **After upgrading** the libraries, you need to **restart the runtime** to make the libraries in sync. 

- Make sure not to execute the cell above (3.1) and below (3.2) again after restarting the runtime.

In [2]:
!pip install -q --upgrade pandas-profiling

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 324.4/324.4 KB 5.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 344.5/344.5 KB 19.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 102.7/102.7 KB 8.9 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 679.5/679.5 KB 13.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 33.8/33.8 MB 13.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.7/4.7 MB 73.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 296.5/296.5 KB 26.2 MB/s eta 0:00:00


# <a name = Section33></a>
### **2.3 Importing Libraries**

In [6]:
import pandas as pd
from pandas_profiling import ProfileReport                          # Importing Pandas Profiling (To generate Univariate Analysis)
pd.set_option('display.max_columns', None)                          # Unfolding hidden features if the cardinality is high
pd.set_option('display.max_colwidth', None)                         # Unfolding the max feature width for better clearity
pd.set_option('display.max_rows', None)                             # Unfolding hidden data points if the cardinality is high
pd.set_option('mode.chained_assignment', None)                      # Removing restriction over chained assignments operations
pd.set_option('display.float_format', lambda x: '%.5f' % x)         # To suppress scientific notation over exponential values
#-------------------------------------------------------------------------------------------------------------------------------                                                 # Importing package numpys (For Numerical Python)
import warnings                                                     # Importing warning to disable runtime warnings
warnings.filterwarnings("ignore")                                   # Warnings will appear only once





    

---

<a name = Section3></a>
# **3. Data Acquisition & Description**

---

In [7]:
df = pd.read_csv('/content/Virat_Kohli_ODI.csv')

print("Data Shape : ", df.shape)

Data Shape :  (132, 12)


### **3.1. Data Description**

In [5]:
df.head(10)

Runs Mins  BF  4s  6s     SR  Pos Dismissal  Inns   Opposition  \
0   12   33  22   1   0  54.54    2       lbw     1  v Sri Lanka   
1   37   82  67   6   0  55.22    2    caught     2  v Sri Lanka   
2   25   40  38   4   0  65.78    1   run out     1  v Sri Lanka   
3   54   87  66   7   0  81.81    1    bowled     1  v Sri Lanka   
4   31   45  46   3   1  67.39    1       lbw     2  v Sri Lanka   
5   2*    6   2   0   0    100    7   not out     1  v Sri Lanka   
6   30   53  41   3   0  73.17    4    caught     2  v Australia   
7   10   32  16   1   0   62.5    3    caught     2  v Australia   
8   27   20  19   4   0  142.1    7    bowled     1  v Sri Lanka   
9   54   93  65   7   0  83.07    4       lbw     1  v Sri Lanka   

          Ground Start Date  
0       Dambulla  18-Aug-08  
1       Dambulla  20-Aug-08  
2  Colombo (RPS)  24-Aug-08  
3  Colombo (RPS)  27-Aug-08  
4  Colombo (RPS)  29-Aug-08  
5  Colombo (RPS)  14-Sep-09  
6       Vadodara  25-Oct-09  
7         Mohali  02-Nov-09  
8         Rajkot  15-Dec-09  
9         Nagpur  18-Dec-09

In [8]:
df.describe()

BF        4s        6s       Pos      Inns
count 132.00000 132.00000 132.00000 132.00000 132.00000
mean   50.87121   4.37121   0.54545   3.30303   1.57576
std    38.72972   4.40403   1.08680   0.87317   0.49611
min     0.00000   0.00000   0.00000   1.00000   1.00000
25%    17.75000   1.00000   0.00000   3.00000   1.00000
50%    42.50000   3.00000   0.00000   3.00000   2.00000
75%    82.25000   7.00000   1.00000   4.00000   2.00000
max   140.00000  18.00000   7.00000   7.00000   2.00000

<a name = Section42></a>
### **3.2 Data Information**

- In this section we will see the **information about the types of features**.

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132 entries, 0 to 131
Data columns (total 12 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Runs        132 non-null    object
 1   Mins        132 non-null    object
 2   BF          132 non-null    int64 
 3   4s          132 non-null    int64 
 4   6s          132 non-null    int64 
 5   SR          132 non-null    object
 6   Pos         132 non-null    int64 
 7   Dismissal   132 non-null    object
 8   Inns        132 non-null    int64 
 9   Opposition  132 non-null    object
 10  Ground      132 non-null    object
 11  Start Date  132 non-null    object
dtypes: int64(5), object(7)
memory usage: 12.5+ KB


---

<a name = Section3></a>
# **4. Data Preprofiling**
---

### **Data Preprofiling Report**

 - **Observations:**

- At some level, we can observe that there are **missing data** in our dataset. 

- Here we can clearly see some **object data type** values, number of **runs scored(Runs), strike rate(SR), minutes(mins)** should be in **integer or float** data type.
- **Date and Time** are found to be **object** rather they should be **DateTime**.

<a name = Section6></a>

---
# **5. Data Cleaning**
---

- In this section, we will perform the **cleaning** operations over the features using information from the previous section.

<a name = Section61></a>
### **5.1 Handling of Inconsitant Data**

In [12]:
df["Runs"] = df["Runs"].str.replace("*", "")
df['Runs']= pd.to_numeric(df['Runs'],errors = 'coerce')
df['Runs']


0       12
1       37
2       25
3       54
4       31
5        2
6       30
7       10
8       27
9       54
10     107
11      91
12     102
13      31
14      57
15       0
16      18
17      10
18      28
19       0
20      37
21     118
22     105
23      64
24      63
25       0
26       2
27      54
28      22
29      28
30      87
31       2
32     100
33      34
34      12
35       1
36      59
37      24
38       9
39      35
40       2
41      81
42       0
43      22
44      94
45       9
46       7
47     107
48      37
49     112
50      35
51      86
52       0
53       3
54     117
55      20
56      23
57      80
58      31
59      18
60      12
61      21
62      66
63     106
64       1
65      38
66     128
67      23
68       0
69       6
70       7
71      15
72      37
73      77
74      26
75       0
76      43
77      11
78     102
79     115
80      14
81      68
82      61
83     100
84      68
85     115
86       0
87      86
88      99
89      19
90      31

<a name = Section611></a>
### **5.2 Handling of Missing Values**

In [13]:
df['Start Date'] = pd.to_datetime(df['Start Date'], utc=True)
df['Mins']= pd.to_numeric(df['Mins'],errors = 'coerce')
df['Mins'].fillna(df['Mins'].mean(), inplace = True)
df['Mins']

0      33.00000
1      82.00000
2      40.00000
3      87.00000
4      45.00000
5       6.00000
6      53.00000
7      32.00000
8      20.00000
9      93.00000
10     70.49206
11    136.00000
12    128.00000
13     70.49206
14     70.49206
15      3.00000
16     28.00000
17     13.00000
18     57.00000
19      5.00000
20     98.00000
21    193.00000
22    162.00000
23     78.00000
24     83.00000
25      3.00000
26     10.00000
27    122.00000
28     43.00000
29     51.00000
30    126.00000
31     11.00000
32    113.00000
33     82.00000
34     16.00000
35      2.00000
36    128.00000
37     38.00000
38     20.00000
39     69.00000
40      5.00000
41    118.00000
42      1.00000
43     68.00000
44    135.00000
45      9.00000
46     18.00000
47    102.00000
48     70.00000
49    134.00000
50     42.00000
51     70.49206
52      5.00000
53     13.00000
54    154.00000
55     21.00000
56     15.00000
57     99.00000
58     30.00000
59     41.00000
60     44.00000
61     36.00000
62    12

In [14]:
df['SR']= pd.to_numeric(df['SR'],errors = 'coerce')
df['SR'].fillna(0, inplace = True)


In [15]:
df.describe()

Runs      Mins        BF        4s        6s        SR       Pos  \
count 132.00000 132.00000 132.00000 132.00000 132.00000 132.00000 132.00000   
mean   46.84848  70.49206  50.87121   4.37121   0.54545  76.99121   3.30303   
std    41.99464  55.94323  38.72972   4.40403   1.08680  37.87449   0.87317   
min     0.00000   1.00000   0.00000   0.00000   0.00000   0.00000   1.00000   
25%    10.00000  19.00000  17.75000   1.00000   0.00000  61.70250   3.00000   
50%    32.50000  65.00000  42.50000   3.00000   0.00000  78.71000   3.00000   
75%    80.25000 120.00000  82.25000   7.00000   1.00000 100.00000   4.00000   
max   154.00000 202.00000 140.00000  18.00000   7.00000 209.09000   7.00000   

           Inns  
count 132.00000  
mean    1.57576  
std     0.49611  
min     1.00000  
25%     1.00000  
50%     2.00000  
75%     2.00000  
max     2.00000

<a name = Section7></a>

---
# **6. Data Post-Profiling**
---

- In this section, we will observe the changes after performing data pre-processing, if present.

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132 entries, 0 to 131
Data columns (total 12 columns):
 #   Column      Non-Null Count  Dtype              
---  ------      --------------  -----              
 0   Runs        132 non-null    int64              
 1   Mins        132 non-null    float64            
 2   BF          132 non-null    int64              
 3   4s          132 non-null    int64              
 4   6s          132 non-null    int64              
 5   SR          132 non-null    float64            
 6   Pos         132 non-null    int64              
 7   Dismissal   132 non-null    object             
 8   Inns        132 non-null    int64              
 9   Opposition  132 non-null    object             
 10  Ground      132 non-null    object             
 11  Start Date  132 non-null    datetime64[ns, UTC]
dtypes: datetime64[ns, UTC](1), float64(2), int64(6), object(3)
memory usage: 12.5+ KB


- In this section we will get **information about the data** and see some observations.

**Observations:**
- The average  **SR(Strike rate)**  wich he has been maintening throughout the years right  **from 2008 till the end of 2017**  is almost  **77**.

- The  **Maximum batting position** where kohli has ever played is **7th**.

- Around **70.50 minutes** of **Average** time he has **spent on ground**, every match he played throughout these 132 matches.

- Around **50.87** average balls he has played in every match **Out of 132 total ODI matches**.

- The **Maximum records for strike rate** and  **Maximum minutes on a match**  out of these 132 ODI matches are found to have **209.09** and **202** respectively.

- The **Maximum number of sixes** and  **Maximum fours scored in a match**  out of 132 ODI matches played are **7 and 18** respectively.

- Out of 132 Matches played between years 2013 to 2017 **virat has played Maximum 24 matches against Sri Lanka** followed by **England and Australia with 21 and 20 number of matches**
 



In [17]:
df['Opposition'].value_counts()

v Sri Lanka       24
v England         21
v Australia       20
v West Indies     19
v South Africa    15
v New Zealand     14
v Bangladesh       8
v Zimbabwe         5
v Pakistan         4
v Ireland          1
v Netherlands      1
Name: Opposition, dtype: int64

In [18]:

df.corr()

Runs     Mins       BF       4s       6s      SR      Pos     Inns
Runs  1.00000  0.93646  0.96164  0.91912  0.57669 0.63506 -0.07077  0.11231
Mins  0.93646  1.00000  0.96125  0.80955  0.48151 0.52738 -0.13748  0.06245
BF    0.96164  0.96125  1.00000  0.85316  0.46487 0.52756 -0.11328  0.10838
4s    0.91912  0.80955  0.85316  1.00000  0.40554 0.63120 -0.08109  0.17046
6s    0.57669  0.48151  0.46487  0.40554  1.00000 0.46622 -0.11116  0.13514
SR    0.63506  0.52738  0.52756  0.63120  0.46622 1.00000  0.19719  0.03626
Pos  -0.07077 -0.13748 -0.11328 -0.08109 -0.11116 0.19719  1.00000 -0.14151
Inns  0.11231  0.06245  0.10838  0.17046  0.13514 0.03626 -0.14151  1.00000

<a name = Section8></a>

---
# **7. Exploratory Data Analysis**
---

In [19]:
import plotly.express as ex 
import numpy as np
import seaborn as sns
import matplotlib.pylab as plt
correlation = df.corr()
fig = ex.imshow(correlation, text_auto=True, aspect="auto")
fig.show()


In [20]:

df['Opposition'].value_counts().index

Index(['v Sri Lanka', 'v England', 'v Australia', 'v West Indies',
       'v South Africa', 'v New Zealand', 'v Bangladesh', 'v Zimbabwe',
       'v Pakistan', 'v Ireland', 'v Netherlands'],
      dtype='object')

### **Question** - How does the opposition country affect the number of runs virat scores?

In [21]:
import plotly.express as ex                                         # Importing Plotly Express for Dynamic Plotting
import plotly.graph_objs as go                                      # Importing Plotly graphs for Dynamic Plotting
import plotly.offline as pyo                                        # Importing offline Express for Dynamic Plotting
from plotly.subplots import make_subplots                           # Importing Plotly Subplots to plot Dynamic subplots
#-------------------------------------------------------------------------------------------------------------------------------
import warnings                                                     # Importing warning to disable runtime warnings
warnings.filterwarnings("ignore")                                   # Warnings will appear only once
#-------------------------------------------------------------------------------------------------------------------------------
from tqdm.notebook import tqdm                                      # To plot progress bar
import plotly.figure_factory as ff                                  # Calling the figure factory to create unique chart types    
#-------------------------------------------------------------------------------------------------------------------------------
import configparser

labels = df['Opposition']
values = df['Runs']

# Making the figure
fig = go.Figure(data=[go.Pie(labels=labels, values=values, hole=.5)])

# Updating the layout
fig.update_layout(height=500, 
                  width=1200, 
                  title_text='ODI Runs Vs Opposition Country', 
                  title_x=0.5)

# Plotting the figure.
fig.show()

In [22]:
labels = df['Dismissal'].value_counts().index
labels

Index(['caught', 'not out', 'run out', 'bowled', 'lbw', 'stumped',
       'hit wicket'],
      dtype='object')

### **Question** - What is the reason & where he gets out very frequently?

In [23]:
labels = df['Dismissal'].value_counts().index
values = df['Dismissal'].value_counts().values

# Making the figure
fig = go.Figure(data=[go.Pie(labels=labels, values=values)])

# Updating the layout
fig.update_layout(height=500, 
                  width=1200, 
                  title_text='Virats ODI dismissals and Reasons', 
                  title_x=0.5) 

# Plotting the figure.
fig.show()

In [24]:

df['Year'] = df['Start Date'].dt.year
df['Year']

0      2008
1      2008
2      2008
3      2008
4      2008
5      2009
6      2009
7      2009
8      2009
9      2009
10     2009
11     2010
12     2010
13     2010
14     2010
15     2010
16     2010
17     2010
18     2010
19     2010
20     2010
21     2010
22     2010
23     2010
24     2010
25     2010
26     2010
27     2011
28     2011
29     2011
30     2011
31     2011
32     2011
33     2011
34     2011
35     2011
36     2011
37     2011
38     2011
39     2011
40     2011
41     2011
42     2011
43     2011
44     2011
45     2011
46     2011
47     2011
48     2011
49     2011
50     2011
51     2011
52     2011
53     2011
54     2011
55     2011
56     2011
57     2011
58     2012
59     2012
60     2012
61     2012
62     2012
63     2012
64     2012
65     2012
66     2012
67     2012
68     2012
69     2013
70     2013
71     2013
72     2013
73     2013
74     2013
75     2013
76     2013
77     2013
78     2013
79     2013
80     2013
81     2013
82     2013
83  

In [33]:
df['Ground'].value_counts()

Dhaka                 7
Dambulla              6
Colombo (RPS)         6
Mohali                6
Kolkata               5
Chennai               5
Delhi                 5
Ahmedabad             4
Harare                4
Visakhapatnam         4
Nagpur                3
Port of Spain         3
Dharamsala            3
Ranchi                3
Bengaluru             3
Rajkot                3
Kochi                 3
Cuttack               3
Sydney                3
Jaipur                3
Melbourne             3
Mumbai                3
Indore                2
Kanpur                2
Brisbane              2
Hambantota            2
Cardiff               2
Hyderabad (Deccan)    2
North Sound           2
Hamilton              2
Vadodara              2
Kingston              2
Pune                  2
Johannesburg          2
Durban                2
Birmingham            2
Napier                1
Wellington            1
Fatullah              1
Nottingham            1
Leeds                 1
Perth           

In [25]:
fig = make_subplots(rows=1, 
                    cols=4, 
                    subplot_titles=('Ground vs Runs', 
                                    'Position vs Runs', 
                                    'Year vs Runs',
                                    'Inning(first/second) vs Runs'))
fig.add_trace(trace=go.Bar(x=df['Ground'], 
                           y=df['Runs'], 
                           orientation='v', 
                           marker=dict(color='rgba(238, 227, 231, 1.0)',
                                       line=dict(color='rgba(148, 42, 43, 1.0)', 
                                                 width=3))),
              row=1, col=1)
fig.add_trace(trace=go.Bar(x=df['Pos'], 
                           y=df['Runs'], 
                           orientation='v', 
                           marker=dict(color='rgba(238, 227, 231, 1.0)',
                                       line=dict(color='rgba(148, 42, 43, 1.0)', 
                                                 width=2))),
              row=1, col=2)
fig.add_trace(trace=go.Bar(x= df['Year'], 
                           y=df['Runs'], 
                           orientation='v', 
                           marker=dict(color='rgba(238, 227, 231, 1.0)',
                                       line=dict(color='rgba(148, 42, 43, 1.0)', 
                                                 width=2))), 
              row=1, col=3)
fig.add_trace(trace=go.Bar(x= df['Inns'], 
                           y=df['Runs'], 
                           orientation='v', 
                           marker=dict(color='rgba(238, 227, 231, 1.0)',
                                       line=dict(color='rgba(148, 42, 43, 1.0)', 
                                                 width=2))), 
              row=1, col=4)
fig.update_layout(height=500, 
                  width=1200, 
                  title_text='Ground vs Runs and Position vs Runs and Year vs Runs and Inning(first/second) vs Runs', 
                  title_x=0.5,  
                  showlegend=False)

# Adding subplot layout axis labels
fig['layout']['xaxis']['title']='Ground'
fig['layout']['yaxis']['title']='Runs'
fig['layout']['xaxis2']['title']='Position'
fig['layout']['yaxis2']['title']='Runs'
fig['layout']['xaxis3']['title']='Year'
fig['layout']['yaxis3']['title']='Runs'
fig['layout']['xaxis4']['title']='Inning(first/second)'
fig['layout']['yaxis4']['title']='Runs'

# Plotting the curve.
fig.show()

<a name = Section9></a>

---
# **9. Summarization**
---

- **<h4>Conclusion</h4>**

    - From this case study it was analyzed that Virat Kohli has **scored most of his ODI carrier runs whenever he batted at 3rd position**.
 
    - He has done very well and have **proven record of performing well while batting in second innings** i.e while chasing.
 
    - As per the data analysis he **has played more number of matches on dhaka ground as compared to any other ground in the world**.
 

 -


-  **<h4>Actionable Insights</h4>**

    - Virat has **higher chances of getting out by getting caught** than any other way like getting stumped or run out.
 
    - He has very **good track record of scoring big runs when chasing and batting at 3rd position than playing at any other batting position** so, there are higher chances of him **scoring more runs when batting at 3rd position while run chasing** .

    - So he should always **preffer himself to bat at 3rd position and focus on not getting out by getting caught**, sometime if needed should go for four instead of six to avoid getting caught.